engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala

```diff
@@ -0,0 +1,102 @@
+import cats.data.NonEmptyList
+import cromwell.core.Dispatcher._
+import cromwell.core._
+import cromwell.engine.workflow.workflowstore.WorkflowStoreActor._
+import cromwell.engine.workflow.workflowstore.WorkflowStoreSubmitActor.{WorkflowSubmittedToStore, WorkflowsBatchSubmittedToStore}
+import cromwell.services.metadata.{MetadataEvent, MetadataKey, MetadataValue}
+import cromwell.services.metadata.MetadataService.PutMetadataAction
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+final case class WorkflowStoreSubmitActor(store: WorkflowStore, serviceRegistryActor: ActorRef) extends Actor with ActorLogging {
+  implicit val ec: ExecutionContext = context.dispatcher
+
+  override def receive = {
+    case cmd: SubmitWorkflow =>
+      val sndr = sender()
+      storeWorkflowSources(NonEmptyList.of(cmd.source)) foreach { ids =>
+        val id = ids.head
+        registerSubmissionWithMetadataService(id, cmd.source)
+        sndr ! WorkflowSubmittedToStore(id)
+        log.info("Workflow {} submitted.", id)
+      }
+    case cmd: BatchSubmitWorkflows =>
+      val sndr = sender()
+      storeWorkflowSources(cmd.sources) foreach { ids =>
+        val assignedSources = ids.toList.zip(cmd.sources.toList)
+        assignedSources foreach { case (id, sourceFiles) => registerSubmissionWithMetadataService(id, sourceFiles) }
+        sndr ! WorkflowsBatchSubmittedToStore(ids)
+        log.info("Workflows {} submitted.", ids.toList.mkString(", "))
+      }
+  }
+
+  private def storeWorkflowSources(sources: NonEmptyList[WorkflowSourceFilesCollection]): Future[NonEmptyList[WorkflowId]] = {
+    for {
+      processedSources <- processSources(sources, _.asPrettyJson)
+      workflowIds <- store.add(processedSources)
+    } yield workflowIds
+  }
+
+  private def processSources(sources: NonEmptyList[WorkflowSourceFilesCollection],
+                             processOptions: WorkflowOptions => WorkflowOptionsJson): Future[NonEmptyList[WorkflowSourceFilesCollection]] = {
+    val nelFutures: NonEmptyList[Future[WorkflowSourceFilesCollection]] = sources map processSource(processOptions)
+    val listFutures: List[Future[WorkflowSourceFilesCollection]] = nelFutures.toList
+    val futureLists: Future[List[WorkflowSourceFilesCollection]] = Future.sequence(listFutures)
+    futureLists.map(seq => NonEmptyList.fromList(seq).get)
+  }
+
+  /**
+    * Runs processing on workflow source files before they are stored.
+    *
+    * @param processOptions How to process the workflow options
+    * @param source         Original workflow source
+    * @return Attempted updated workflow source
+    */
+  private def processSource(processOptions: WorkflowOptions => WorkflowOptionsJson)
+                           (source: WorkflowSourceFilesCollection): Future[WorkflowSourceFilesCollection] = {
+    val options = Future {
+      WorkflowOptions.fromJsonString(source.workflowOptionsJson)
```

Comment:
TOL: Presumably carry-over but I didn't think we even validated that it was JSON until the WFMaterializationActor, so that we get an ID and a record for every submission, even if it's invalid?